### DISCUSSION TITLE
[v5] How to use dto conveniently?

### DISCUSSION THREAD

**Discussion by AGDholo** [adonisjs/core#1797] [Help]
Our team try to use dto in adonisjs now,  
But there are some problems.

1. We export a dto class with model:

```
export class GameResponse implements Identity, EventTime {
  public id: number
  public appid: number
  public name: string
  public createdAt: DateTime
  public updatedAt: DateTime
}
```

and push query res data into the class every time:

```
    await Game.query()
      .orderBy('created_at', 'desc')
      .paginate(requestGet.page, requestGet.limit)
      .then(item => {
        item.forEach(f => {
          result.push({
            id: f.id,
            appid: f.appid,
            name: f.name,
            createdAt: f.createdAt,
            updatedAt: f.updatedAt,
          })
        })
      })
    return result
```

This is very tedious and pushes repeatedly for each query.

2. Cannot parse the paginate(meta data and so on) easily.

We open-source the code now:

* Controller(ORM query): https://github.com/OpenEpicData/EpicDataAPI/blob/develop/app/Controllers/Http/GamesController.ts
* DTO: https://github.com/OpenEpicData/EpicDataAPI/blob/develop/app/Middleware/Models/Game.ts

**Comment by Banashek** [comment]
This is how I might write the controller endpoint to utilize the explicit DTO:

```ts
public async index ({ request }): Promise<GameResponse[]> {
  return await Game.query()
    .orderBy('created_at', 'desc')
    .paginate(requestGet.page, requestGet.limit)
    .select('id', 'appid', 'name', 'createdAt', 'updatedAt');
}
```

Not 100% on if there will be type errors there, but by declaring the return type of the function to be of your DTO type, you'll get type checking on that. Then, (and this is the part I'm not 100% on) if you specify only the columns of your DTO to select from the query, it _should_ be able to infer the return type.
